trl/environment/base.py

```diff
@@ -89,32 +91,78 @@ def __init__(self, model, tokenizer, tools, reward_fn, prompt, max_turns=4, gene
             self.generation_kwargs = generation_kwargs
+        self.generation_kwargs["pad_token_id"] = tokenizer.eos_token_id
 
         self.is_encoder_decoder = hasattr(self.model, "is_encoder_decoder")
         self.current_device = extract_model_from_parallel(self.model).pretrained_model.device
 
+    def encode_history(self, history, is_system_prompt=False):
+        encoded_prompt = self.tokenizer(history.last_text_segment, return_tensors="pt").to(self.current_device)
+        input_ids = encoded_prompt["input_ids"]
+
+        if is_system_prompt:
+            attention_mask = torch.zeros_like(input_ids)
+        else:
+            attention_mask = torch.ones_like(input_ids)
+
+        return {"input_ids": input_ids, "attention_mask": attention_mask}
+
     def run(self, tasks):
         turns = 0
         histories = [TextHistory(self.prompt + task, system=True) for task in tasks]
 
+        encoded_prompts = [
+            self.tokenizer(self.prompt + task, return_tensors="pt").to(self.current_device) for task in tasks
+        ]
+
+        # queries
+        histories_queries = [encoded_prompt["input_ids"].squeeze() for encoded_prompt in encoded_prompts]
+        queries_masks = [encoded_prompt["attention_mask"].squeeze() for encoded_prompt in encoded_prompts]
+
+        # respo,ses
+        histories_responses = [[] for _ in range(len(tasks))]
+        responses_masks = [[] for _ in range(len(tasks))]
+
+        # rewards
+        histories_rewards = []
+
         while any([not history.completed for history in histories]):
             histories = self.generate(histories)
+
             # TODO: make this parallel rather than for-loop
             for i in range(len(histories)):
+                encoded_history = self.encode_history(histories[i])
+                histories_responses[i].append(encoded_history["input_ids"].squeeze())
+                responses_masks[i].append(encoded_history["attention_mask"].squeeze())
+
                 histories[i] = self.step(histories[i])
+
+                is_system_prompt = not histories[i].completed
+                encoded_history = self.encode_history(histories[i], is_system_prompt=is_system_prompt)
+
+                histories_responses[i].append(encoded_history["input_ids"].squeeze())
+                responses_masks[i].append(encoded_history["attention_mask"].squeeze())
+
             turns += 1
             if turns == self.max_turns:
                 break
 
-        self.compute_reward(histories)
+        histories = self.compute_reward(histories)
 
-        return histories
+        for i, history in enumerate(histories):
+            histories_rewards.append(torch.Tensor([history.reward]))
+            histories_responses[i] = torch.cat(histories_responses[i], dim=0).squeeze()
+            responses_masks[i] = torch.cat(responses_masks[i], dim=0).squeeze()
+
+        return (histories_queries, histories_responses), (responses_masks, queries_masks), histories_rewards, histories
```

Comment:
Still not a big fan of the nested returns, I think this will make it not very user friendly.